### DISCUSSION TITLE
TestStore not receiving effect when used as stored property in test classes

### DISCUSSION THREAD

**Discussion by adityadaniel** [pointfreeco/swift-composable-architecture#3052] [Q&A]
Hi @mbrandonw @stephencelis, My teammate found surprising behaviour when working on unit test that has asynchronous / long living effect. The test setup can be described as follow (I use Effect basic tests from SwiftUI case studies)
```swift
final class EffectsBasicsTests: XCTestCase {
  private let testStore = TestStore(initialState: EffectsBasics.State()) {
      EffectsBasics()
    } withDependencies: {
      $0.factClient.fetch = { "\($0) is a good number Brent" }
      $0.continuousClock = ImmediateClock()
    }

  @MainActor
  func testNumberFact() async {
    await testStore.send(.incrementButtonTapped) {
      $0.count = 1
    }
    await testStore.send(.numberFactButtonTapped) {
      $0.isNumberFactRequestInFlight = true
    }
//    await testStore.receive(\.numberFactResponse.success) {
//      $0.isNumberFactRequestInFlight = false
//      $0.numberFact = "1 is a good number Brent"
//    }
  }
}
```

Instead of creating TestStore in each test cases, we create TestStore as stored property and reusing them between each test cases. If you run test case above, it will succeed. There's no test failure on not receiving effect.

However when we follow the test example by creating TestStore locally on each test case function, the test will failed on not receiving effect. Which we expect that. 

Is creating TestStore as stored property not supported in TCA and is it expected behaviour? Does creating TestStore locally on each test case function the best practice on writing unit test?

**Comment by Matejkob** [comment]
Hi @adityadaniel ,

The `TestStore` is designed to provide an exhaustive testing environment where each test is expected to run in a completely isolated context. This isolation helps ensure that the state and effects from one test do not inadvertently influence another. <s>When `TestStore` is used as a stored property, the isolation between tests can be compromised, as the state and the completion of effects might carry over from one test to another. </s> Furthermore, [the deinitialization process](https://github.com/pointfreeco/swift-composable-architecture/blob/c0a694451f15446c88058bb8c5bbfb1a78b22f81/Sources/ComposableArchitecture/TestStore.swift#L614) of the `TestStore`, which performs a final check on unreceived actions and unfinished effects, will not be executed when the `TestStore` is not reinitialized for each test. This can potentially mask issues that would otherwise cause a test to fail, providing a false sense of correctness in the test suite.

Here’s a more appropriate setup that aligns with the principles of isolated and exhaustive testing:

```swift
final class EffectsBasicsTests: XCTestCase {
  private var testStore: TestStore<EffectsBasics.State, EffectsBasics.Action>!
  
  override func setUp() {
    super.setUp()
    testStore = TestStore(initialState: EffectsBasics.State()) {
      EffectsBasics()
    } withDependencies: {
      $0.factClient.fetch = { "\($0) is a good number Brent" }
      $0.continuousClock = ImmediateClock()
    }
  }
  
  override func tearDown() {
    testStore = nil
    super.tearDown()
  }
  
  @MainActor
  func testNumberFact() async {
    await testStore.send(.incrementButtonTapped) {
      $0.count = 1
    }
    await testStore.send(.numberFactButtonTapped) {
      $0.isNumberFactRequestInFlight = true
    }
    // This would be where you expect the effect's response
    // await testStore.receive(.numberFactResponse.success("1 is a good number Brent")) {
    //   $0.isNumberFactRequestInFlight = false
    //   $0.numberFact = "1 is a good number Brent"
    // }
  }
}
```

<s>This setup ensures that each test begins with a fresh `TestStore`, thereby maintaining the integrity and isolation of the tests.</s> It allows the `deinit` of `TestStore` to perform necessary checks for unreceived actions and lingering effects, which are crucial for ensuring the accuracy and completeness of your tests.

**Reply by stephencelis** [reply]
We've also added #3054 to make asserting in this fashion a bit easier using `store.finish()`.

**Reply by adityadaniel** [reply]
Hi @Matejkob thank you for thorough explanation! I still have some questions.

> When TestStore is used as a stored property, the isolation between tests can be compromised, as the state and the completion of effects might carry over from one test to another.

Does it means that the state and effects could be affected by another test cases on the same class? When I try to debug this, I observe that `self.reducer.receivedActions.count` return 0 instead of some number in the `deinit`.

**Reply by stephencelis** [reply]
@adityadaniel I think @Matejkob may be mistaken here. XCTest runs each test function in isolation, so each technically would get its own `TestStore`.

**Reply by Matejkob** [reply]
Apology; my mistake. @stephencelis is right that, XCTest indeed runs each test function in isolation.

**Comment by stephencelis** [comment]
@adityadaniel In general we do not recommend holding onto the test store in this fashion. @Matejkob does a good job breaking down why you're seeing the behavior you're seeing, but also we just don't think initializing the test store at the level of the test class is a good idea.

In general, each test should be responsible for initializing a feature's state and dependencies for what is being tested. A lot of the time you will want to provide different initial state or dependencies depending on what you are testing, and so initializing the test store the same way each time unfortunately goes against this.

We'll be adding docs for this in #3054, which also includes a helper that at least allows you to get exhaustive test coverage back in a manual way, but we highly recommend initializing test stores in each individual test, instead.

**Comment by adityadaniel** [comment]
Tagging @jeffersonsetiawan @wendyliga @dikasetiadi for visibility
